lib/TH/THGeneral.c

```diff
@@ -139,17 +142,23 @@ static long getAllocSize(void *ptr) {
  *     soft max by 40%
  */
 static void maybeTriggerGC() {
-  if(torchGCFunction && torchHeapSize > torchHeapSizeSoftMax) {
+  if(torchGCFunction && heapSize > heapSoftmax ) {
     torchGCFunction(torchGCData);
-    if(torchHeapSize > torchHeapSizeSoftMax * 0.8) {
-      torchHeapSizeSoftMax = torchHeapSizeSoftMax * 1.4;
+    if(heapSize > heapSoftmax * heapSoftmaxGrowthThresh) {
+      heapSoftmax = heapSoftmax * heapSoftmaxGrowthFactor;
     }
   }
 }
 
 // hooks into the TH heap tracking
 void THHeapUpdate(long size) {
-  torchHeapSize += size;
+  THAtomicAddLong(&heapSize, size);
+
+# ifdef TH_CHECK_HEAP_UPDATE
+  if (heapSize < 0)
```

Comment:
I think you're right, although I'm not sure I agree with your justification.

The atomic add will introduce a memory fence, so I think `heapSize` will have seen all prior updates at that point. However, if a thread writes `heapSize` after the add, the read could see tearing. I'll make sure I read `heapSize` atomically, thanks.